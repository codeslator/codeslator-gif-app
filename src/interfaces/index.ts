import { Dispatch, SetStateAction } from 'react'

export interface AddCategoryProps {
  categories: string[],
  setCategories: Dispatch<SetStateAction<string[]>>
}

export type GifGridProps = {
  category: string;
}