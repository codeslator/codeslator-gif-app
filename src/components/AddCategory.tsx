import { useState, FormEvent, SyntheticEvent, FC } from 'react';
import { AddCategoryProps } from '../interfaces/index';

export const AddCategory: FC<AddCategoryProps> = ({ categories, setCategories }) => {
  const [value, setValue] = useState<string>('');

  const handleInput = (event: FormEvent<HTMLInputElement>) => {
    setValue(event.currentTarget.value)
  };

  const handleSubmit = (e: SyntheticEvent) => {
    e.preventDefault();
    if(value !== '') {
      setCategories((prev: string[]) => [...prev, value]);
      setValue('');
    }
  };

  return (
    <>
      <form onSubmit={handleSubmit}>
        <h2>Search Category</h2>
        <div className="row">
          <div className="col-9">
            <input
              type="text"
              className="form-control"
              placeholder="Category"
              id="category"
              value={value}
              onChange={handleInput}
            />
          </div>
          <div className="col-3">
            <button className="btn btn-primary btn-full" type="submit">Submit</button>
          </div>
        </div>
      </form>
      <p>{value}</p>
    </>
  )
}
