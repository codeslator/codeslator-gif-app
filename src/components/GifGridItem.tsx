import { GifGridItemProps } from '../interfaces/index';

export const GifGridItem = ({ gif }: GifGridItemProps) => {
  const { title, url } = gif;
  return (
    <div className="card mb-3 animate__animated animate__fadeIn">
      <img src={url} className="card-img-top" alt={title} />
      <div className="card-body">
        <h5 className="card-title">{title}</h5>
      </div>
    </div>
  )
}
