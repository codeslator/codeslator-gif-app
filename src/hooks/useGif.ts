import { useState } from 'react';
import { gifApi } from '../apis/gifApi';
import { GifData } from '../interfaces/index';

const initialState = {
  gifs: [],
  isLoading: true,
}

const useGif = () => {
  const [state, setState] = useState<GifData>(initialState);

  const fetchGifs = async (search: string) => {
    const { data } = await gifApi.get('', {
      params: {
        q: encodeURI(search)
      }
    });
    const gifs = data.data.map((img: any) => {
      return {
        id: img.id,
        title: img.title,
        url: img.images?.downsized_medium.url
      };
    });
    setState({ gifs, isLoading: false });
  }

  return {
    gifs: state.gifs,
    isLoading: state.isLoading,
    fetchGifs,
  };
};

export default useGif;