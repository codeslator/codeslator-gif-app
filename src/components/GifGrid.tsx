import { FC, useEffect } from 'react';
import { GifGridProps } from '../interfaces/index';
import useGif from '../hooks/useGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid: FC<GifGridProps> = ({ category }) => {
  const { fetchGifs, gifs, isLoading } = useGif();

  useEffect(() => {
    fetchGifs(category)
  }, [category, fetchGifs])

  return (
    <>
      <h3 className="mb-3">{category}</h3>
      <div className="row">
        {isLoading && (
          <h4 className="text-center">Loading...</h4>
        )}
        {gifs.map((gif) => (
          <div className="col-xs-6 col-sm-4 col-md-3 col-lg-2" key={gif.id} >
            <GifGridItem gif={gif} />
          </div>
        ))}
      </div>
    </>
  )
}
