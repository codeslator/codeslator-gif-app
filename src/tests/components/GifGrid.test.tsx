import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import { GifGrid } from '../../components/GifGrid';
import useGif from '../../hooks/useGif';
import { Gif } from '../../interfaces';
jest.mock('../../hooks/useGif');

describe('GifGrid Component', () => {
  
  test('Should to show the component correctly', () => {
    useGif.mockReturnValue({
      gifs: [],
      isLoading: true,
      fetchGifs: jest.fn()
    })
    const category = 'Hello World';
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Should to show items when images are loaded', () => {
    const gifs: Gif[] = [
      {
        id: '1',
        title: 'Hello World',
        url: 'http://localhost:3000/image.png'
      },
      {
        id: '2',
        title: 'Hello World 1',
        url: 'http://localhost:3000/image1.png'
      },
    ]
    useGif.mockReturnValue({
      gifs,
      isLoading: false,
      fetchGifs: jest.fn()
    })
    const category = 'Hello World';
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h4').exists()).toBeFalsy();
    expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
  })
})