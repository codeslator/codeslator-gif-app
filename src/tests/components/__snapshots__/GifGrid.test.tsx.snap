// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid Component Should to show items when images are loaded 1`] = `
<Fragment>
  <h3
    className="mb-3"
  >
    Hello World
  </h3>
  <div
    className="row"
  >
    <div
      className="col-xs-6 col-sm-4 col-md-3 col-lg-2"
      key="1"
    >
      <GifGridItem
        gif={
          Object {
            "id": "1",
            "title": "Hello World",
            "url": "http://localhost:3000/image.png",
          }
        }
      />
    </div>
    <div
      className="col-xs-6 col-sm-4 col-md-3 col-lg-2"
      key="2"
    >
      <GifGridItem
        gif={
          Object {
            "id": "2",
            "title": "Hello World 1",
            "url": "http://localhost:3000/image1.png",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`GifGrid Component Should to show the component correctly 1`] = `
<Fragment>
  <h3
    className="mb-3"
  >
    Hello World
  </h3>
  <div
    className="row"
  >
    <h4
      className="text-center"
    >
      Loading...
    </h4>
  </div>
</Fragment>
`;
